core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java

```diff
@@ -146,36 +153,41 @@ public URI getLocation() {
   /**
    * @param entry an entry to write to the journal
    */
-  public void write(JournalEntry entry) throws IOException {
+  @VisibleForTesting
+  synchronized void write(JournalEntry entry) throws IOException {
     writer().write(entry);
   }
 
   /**
    * Flushes the journal.
    */
-  public void flush() throws IOException {
+  @VisibleForTesting
+  synchronized void flush() throws IOException {
     writer().flush();
   }
 
   @Override
-  public JournalContext createJournalContext() {
-    if (mAsyncWriter == null) {
-      return new NoopJournalContext();
+  public synchronized JournalContext createJournalContext() throws UnavailableException {
+    if (mState == State.CLOSED) {
+      // We throw UnavailableException here so that clients will retry with the next primary master.
+      throw new UnavailableException("Failed to write to journal: journal is closed");
```

Comment:
Should we throw this if the state is secondary?

Reply:
done